rasterio/_base.pyx

```diff
@@ -449,6 +450,63 @@ cdef class DatasetBase(object):
                     GDALGetRasterUnitType(self.band(j)) for j in self.indexes)
             return self._units
 
+
+    def block_window(self, bidx, i, j):
+        """Returns the window for a particular block
```

Comment:
@perrygeo a complication of the previous `block()` method was that it could partially succeed and partially fail. I think we should eliminate that complexity by splitting it into two methods that can each only pass or fail.